reactioncommerce/reaction
# Replace CFS

Resolves #1221   
Impact: breaking 
Type: feature

## Issue

The Meteor CollectionFS packages are not maintained. Although they work fine in most cases, we have limited ability to troubleshoot and fix bugs, and we can't easily implement additional stores and file types.

## Solution

The Meteor CollectionFS packages are gone and are replaced with new, similar but cleaner and more modern NPM packages. These live here: https://github.com/reactioncommerce/reaction-file-collections

- The main client and server files for file handling are now in /imports/plugins/core/files
- There is now a `MediaRecords` backing collection exported with the other collections, and a `Media` MeteorFileCollection exported from the files plugin. `MediaRecords` is the equivalent of what you previously got from doing `Media.files`, i.e., the direct collection reference. But going forward it should just be imported like any other collection.
- I've consolidated some of the duplicated code around getting product images (see `getPrimaryMediaForItem`). Also `Reaction.Email.getShopLogo` for emails. There may even be places where more reuse could happen, but sometimes the Media selector is slightly different and we'd have to think about whether those differences are truly necessary.
- I created a quick React component to replace the brand image uploader/selector in Shop Settings. This probably needs some design love, but it's functional. We may want to tackle better design for it as a separate ticket.
- I did NOT remove MediaRecords collection hooks. Will create a separate ticket for that.
- Added indexes on MediaRecords collection to make those queries faster.

## Breaking changes

If you've saved the file URLs anywhere, they're now different.

```
/assets/files/:collectionName/:fileId/:filename
```

becomes

```
/assets/files/:collectionName/:fileId/:primaryStoreName/:filename
```

and

```
/assets/files/:collectionName/:fileId/:filename?store=storeName
```

becomes

```
/assets/files/:collectionName/:fileId/:storeName/:filename
```

I've deleted some unused Blaze templates rather than update URL handling within them:

- shopBrandImageOption
- ordersListItems
- select
- upload
- productMetaField
- productMetaFieldForm
- metaComponent
- productDetailEdit
- productDetailField
- productImageGallery
- imageDetail
- imageUploader
- productSocial
- variantList
- variant

Media-related publishing is changed and improved:

- "CartItemImage" publication is removed
- "CartImages" now takes an ID
- Added "ProductGridMedia" to replace Media being included with the products publication for the grid
- Added "ProductMedia"
- Added "OrderImages", similar to CartImages, used for order now rather than reusing CartImages


## Testing

Test each of these
- logged in as admin
- logged in as non-admin
- not logged in

1. Grid - Product images show up and look correct for each different view option
1. PDP - Product images look correct
1. Emails
    - make sure the correct shop logo is used in all emails, or the correct fallback logo if there isn't one

And test these only as admin:

1. Grid - thumbnail image in the panel appears
1. Shop settings
    - Click to add brand images
    - Drop brand images
    - Click brand image to change
    - Delete brand images (and if you cancel, make sure not deleted)
    - Try to upload non-images (shouldn't be allowed)
1. PDP
    - Click + to add images
    - Drop images
    - Try to upload non-images (shouldn't be allowed)
    - Drag to reorder images
    - Zoom view should work when not an admin
    - Delete product images (and if you cancel, make sure not deleted)
1. Orders
    - Make sure correct photo appears next to each product in invoice
1. Cloning products with media
    - make sure it succeeds and all images are also copied, in all stores 
1. Shopify Import
    - Make sure product images come over
1. Anywhere else that you see was touched when reviewing the code diff
